{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gevers et al Case Study \n",
    "\n",
    "In this notebook, we will take concepts we learned about in `the-importance-of-being-scale-invariant.ipynb` in an Inflammatory Bowel Disease (IBD) microbiome dataset.  In particular, we will use a subset of the [Gevers et al](https://pubmed.ncbi.nlm.nih.gov/24629344/) dataset that was retrieved from [Qiita](https://qiita.ucsd.edu/study/description/1939)\n",
    "\n",
    "\n",
    "The goal of this notebook is to (1) determine if there is a difference between patients who have IBD and those that don't and (2) determine which microbes could explain these differences. The broader framework we will be exploring is called Differential Abundance analysis (or Differential Expression analysis if you analyzing RNAseq data, the concept is identical). To answer these questions, we will use a probabilistic programming framework called Stan to allow us to fit a Generalized Linear Model on the Gevers et al dataset.  Specifically, we will fit a Negative Binomial regression using the ALR transform that we covered earlier.\n",
    "\n",
    "One may ask, why would we want to go Bayesian?\n",
    "\n",
    "There are a number of reasons why we would want to employ a Bayesian approach to differential abundance analysis, namely\n",
    "1. Our dataset is highly sparse.  This means that if we want to employ the log-ratio methodology discussed earlier, the Differential Abundance problem becomes a missing data problem, since logarithms cannot be evaluated on zeros.  This approach quickly becomes ill-defined in a Frequentist regime (i.e. how would one deal with microbes that are only found in 1 group?).\n",
    "2. We will want to employ hierarchical models that can explicitly model technical variation induced by counting sequences, while keeping the model parameters interpretable in the compositional data analysis framework we outlined earlier.\n",
    "\n",
    "The Negative Binomial regression framework is an example of a simple hierarchical model that is already commonly used by tools such as edgeR and DESeq2.  Here, we will show you how to design your own model in Stan and interpret the model quantities.\n",
    "\n",
    "But first, we'll need to load the appropriate Python libraries and datasets.  In particular, we will want to load a table of the microbial counts collected across samples, in addition to the accomodating sample metadata information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cmdstanpy import CmdStanModel\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "%matplotlib inline\n",
    "\n",
    "md = pd.read_table('../data/gevers/rectum_metadata.txt', index_col=0)\n",
    "table = pd.read_table('../data/gevers/rectum_table.txt', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The microbial count table (commonly referred to as the Biom Table), is a profile of microbial \"species\" counts within each sample. The reason why \"species\" is highlighted is because the term microbial species is still not standardized -- unlike Eurkaryotic organisms, microbes can exchange DNA amongst completely unrelated microbes, defying the standard definitions of species.  Because of this edge case, there are a number of methodologies that have attempted to carve out their alternative to microbial species, commonly called OTUs (Operational Taxonomic Units).  \n",
    "\n",
    "In this particular case, we have collapsed the OTUs to Genus level, where OTU abundances within each genus are averaged together.  This allows us to simplify the dataset reducing the dimensionality from 8000+ OTUs to 113 genera.  As a result, we have reduced the taxonomic resolution, but the resulting data is easier to handle and easier to interpret.  Below is an example of what this dataset would look like, where the column names correspond to the microbial taxonomic names and the rownames correspond to biological samples collected from rectum swabs.  The values in below represent the number of times a genus was observed in the corresponding biological sample using 16S sequencing.  With 16S sequencing, 1 single sequencing read is enough to uniquely identify 1 microbe, so we can just count the number of sequencing reads to count the number of microbes.  There are certainly edge cases to this procedure, but that is a discussion for another time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Actinomycetaceae;g__Actinomyces</th>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Corynebacteriaceae;g__Corynebacterium</th>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Micrococcaceae;g__Rothia</th>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium</th>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__</th>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Adlercreutzia</th>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Atopobium</th>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Collinsella</th>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Eggerthella</th>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Slackia</th>\n",
       "      <th>...</th>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pasteurellales;f__Pasteurellaceae;g__Actinobacillus</th>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pasteurellales;f__Pasteurellaceae;g__Aggregatibacter</th>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pasteurellales;f__Pasteurellaceae;g__Haemophilus</th>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Moraxellaceae;g__Acinetobacter</th>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Moraxellaceae;g__Moraxella</th>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Pseudomonadaceae;g__</th>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Pseudomonadaceae;g__Pseudomonas</th>\n",
       "      <th>k__Bacteria;p__Spirochaetes;c__Spirochaetes;o__Spirochaetales;f__Spirochaetaceae;g__Treponema</th>\n",
       "      <th>k__Bacteria;p__Tenericutes;c__Mollicutes;o__RF39;f__;g__</th>\n",
       "      <th>k__Bacteria;p__Verrucomicrobia;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Verrucomicrobiaceae;g__Akkermansia</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1939.MGH100695.a</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939.MGH100695.b</th>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>185.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939.MGH101446</th>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>179.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>249.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2280.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939.MGH101666</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939.MGH102261</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 113 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Actinomycetaceae;g__Actinomyces  \\\n",
       "sample_id                                                                                                                 \n",
       "1939.MGH100695.a                                                0.0                                                       \n",
       "1939.MGH100695.b                                               39.0                                                       \n",
       "1939.MGH101446                                                 10.0                                                       \n",
       "1939.MGH101666                                                  0.0                                                       \n",
       "1939.MGH102261                                                  0.0                                                       \n",
       "\n",
       "                  k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Corynebacteriaceae;g__Corynebacterium  \\\n",
       "sample_id                                                                                                                       \n",
       "1939.MGH100695.a                                                0.0                                                             \n",
       "1939.MGH100695.b                                                0.0                                                             \n",
       "1939.MGH101446                                                  0.0                                                             \n",
       "1939.MGH101666                                                  1.0                                                             \n",
       "1939.MGH102261                                                  0.0                                                             \n",
       "\n",
       "                  k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Micrococcaceae;g__Rothia  \\\n",
       "sample_id                                                                                                          \n",
       "1939.MGH100695.a                                                0.0                                                \n",
       "1939.MGH100695.b                                                0.0                                                \n",
       "1939.MGH101446                                                  0.0                                                \n",
       "1939.MGH101666                                                  1.0                                                \n",
       "1939.MGH102261                                                  0.0                                                \n",
       "\n",
       "                  k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium  \\\n",
       "sample_id                                                                                                                         \n",
       "1939.MGH100695.a                                                0.0                                                               \n",
       "1939.MGH100695.b                                               38.0                                                               \n",
       "1939.MGH101446                                                  3.0                                                               \n",
       "1939.MGH101666                                                  1.0                                                               \n",
       "1939.MGH102261                                                  1.0                                                               \n",
       "\n",
       "                  k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__  \\\n",
       "sample_id                                                                                                        \n",
       "1939.MGH100695.a                                                0.0                                              \n",
       "1939.MGH100695.b                                                9.0                                              \n",
       "1939.MGH101446                                                  3.0                                              \n",
       "1939.MGH101666                                                  0.0                                              \n",
       "1939.MGH102261                                                  0.0                                              \n",
       "\n",
       "                  k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Adlercreutzia  \\\n",
       "sample_id                                                                                                                     \n",
       "1939.MGH100695.a                                                0.0                                                           \n",
       "1939.MGH100695.b                                                0.0                                                           \n",
       "1939.MGH101446                                                  1.0                                                           \n",
       "1939.MGH101666                                                  0.0                                                           \n",
       "1939.MGH102261                                                  0.0                                                           \n",
       "\n",
       "                  k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Atopobium  \\\n",
       "sample_id                                                                                                                 \n",
       "1939.MGH100695.a                                                0.0                                                       \n",
       "1939.MGH100695.b                                                0.0                                                       \n",
       "1939.MGH101446                                                  8.0                                                       \n",
       "1939.MGH101666                                                  0.0                                                       \n",
       "1939.MGH102261                                                  0.0                                                       \n",
       "\n",
       "                  k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Collinsella  \\\n",
       "sample_id                                                                                                                   \n",
       "1939.MGH100695.a                                                7.0                                                         \n",
       "1939.MGH100695.b                                               53.0                                                         \n",
       "1939.MGH101446                                                179.0                                                         \n",
       "1939.MGH101666                                                  0.0                                                         \n",
       "1939.MGH102261                                                  0.0                                                         \n",
       "\n",
       "                  k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Eggerthella  \\\n",
       "sample_id                                                                                                                   \n",
       "1939.MGH100695.a                                                0.0                                                         \n",
       "1939.MGH100695.b                                                0.0                                                         \n",
       "1939.MGH101446                                                  0.0                                                         \n",
       "1939.MGH101666                                                  0.0                                                         \n",
       "1939.MGH102261                                                  1.0                                                         \n",
       "\n",
       "                  k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Slackia  \\\n",
       "sample_id                                                                                                               \n",
       "1939.MGH100695.a                                                0.0                                                     \n",
       "1939.MGH100695.b                                                9.0                                                     \n",
       "1939.MGH101446                                                  0.0                                                     \n",
       "1939.MGH101666                                                  0.0                                                     \n",
       "1939.MGH102261                                                  0.0                                                     \n",
       "\n",
       "                  ...  \\\n",
       "sample_id         ...   \n",
       "1939.MGH100695.a  ...   \n",
       "1939.MGH100695.b  ...   \n",
       "1939.MGH101446    ...   \n",
       "1939.MGH101666    ...   \n",
       "1939.MGH102261    ...   \n",
       "\n",
       "                  k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pasteurellales;f__Pasteurellaceae;g__Actinobacillus  \\\n",
       "sample_id                                                                                                                       \n",
       "1939.MGH100695.a                                                4.0                                                             \n",
       "1939.MGH100695.b                                                0.0                                                             \n",
       "1939.MGH101446                                                  0.0                                                             \n",
       "1939.MGH101666                                                  0.0                                                             \n",
       "1939.MGH102261                                                  0.0                                                             \n",
       "\n",
       "                  k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pasteurellales;f__Pasteurellaceae;g__Aggregatibacter  \\\n",
       "sample_id                                                                                                                        \n",
       "1939.MGH100695.a                                                0.0                                                              \n",
       "1939.MGH100695.b                                                4.0                                                              \n",
       "1939.MGH101446                                                249.0                                                              \n",
       "1939.MGH101666                                                 10.0                                                              \n",
       "1939.MGH102261                                                  8.0                                                              \n",
       "\n",
       "                  k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pasteurellales;f__Pasteurellaceae;g__Haemophilus  \\\n",
       "sample_id                                                                                                                    \n",
       "1939.MGH100695.a                                                4.0                                                          \n",
       "1939.MGH100695.b                                              130.0                                                          \n",
       "1939.MGH101446                                                 44.0                                                          \n",
       "1939.MGH101666                                                221.0                                                          \n",
       "1939.MGH102261                                                  0.0                                                          \n",
       "\n",
       "                  k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Moraxellaceae;g__Acinetobacter  \\\n",
       "sample_id                                                                                                                     \n",
       "1939.MGH100695.a                                                0.0                                                           \n",
       "1939.MGH100695.b                                               58.0                                                           \n",
       "1939.MGH101446                                                  0.0                                                           \n",
       "1939.MGH101666                                                  0.0                                                           \n",
       "1939.MGH102261                                                 13.0                                                           \n",
       "\n",
       "                  k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Moraxellaceae;g__Moraxella  \\\n",
       "sample_id                                                                                                                 \n",
       "1939.MGH100695.a                                                0.0                                                       \n",
       "1939.MGH100695.b                                                0.0                                                       \n",
       "1939.MGH101446                                                  0.0                                                       \n",
       "1939.MGH101666                                                  0.0                                                       \n",
       "1939.MGH102261                                                  0.0                                                       \n",
       "\n",
       "                  k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Pseudomonadaceae;g__  \\\n",
       "sample_id                                                                                                           \n",
       "1939.MGH100695.a                                                0.0                                                 \n",
       "1939.MGH100695.b                                                0.0                                                 \n",
       "1939.MGH101446                                                  0.0                                                 \n",
       "1939.MGH101666                                                  0.0                                                 \n",
       "1939.MGH102261                                                  0.0                                                 \n",
       "\n",
       "                  k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Pseudomonadaceae;g__Pseudomonas  \\\n",
       "sample_id                                                                                                                      \n",
       "1939.MGH100695.a                                                0.0                                                            \n",
       "1939.MGH100695.b                                                0.0                                                            \n",
       "1939.MGH101446                                                  0.0                                                            \n",
       "1939.MGH101666                                                  1.0                                                            \n",
       "1939.MGH102261                                                  0.0                                                            \n",
       "\n",
       "                  k__Bacteria;p__Spirochaetes;c__Spirochaetes;o__Spirochaetales;f__Spirochaetaceae;g__Treponema  \\\n",
       "sample_id                                                                                                         \n",
       "1939.MGH100695.a                                                0.0                                               \n",
       "1939.MGH100695.b                                                0.0                                               \n",
       "1939.MGH101446                                                  0.0                                               \n",
       "1939.MGH101666                                                  0.0                                               \n",
       "1939.MGH102261                                                  0.0                                               \n",
       "\n",
       "                  k__Bacteria;p__Tenericutes;c__Mollicutes;o__RF39;f__;g__  \\\n",
       "sample_id                                                                    \n",
       "1939.MGH100695.a                                                0.0          \n",
       "1939.MGH100695.b                                              185.0          \n",
       "1939.MGH101446                                                  5.0          \n",
       "1939.MGH101666                                                  0.0          \n",
       "1939.MGH102261                                                  0.0          \n",
       "\n",
       "                  k__Bacteria;p__Verrucomicrobia;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Verrucomicrobiaceae;g__Akkermansia  \n",
       "sample_id                                                                                                                         \n",
       "1939.MGH100695.a                                                0.0                                                               \n",
       "1939.MGH100695.b                                                3.0                                                               \n",
       "1939.MGH101446                                               2280.0                                                               \n",
       "1939.MGH101666                                                 28.0                                                               \n",
       "1939.MGH102261                                                  0.0                                                               \n",
       "\n",
       "[5 rows x 113 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sample metadata information provides information surrounding the sample collection in addition to other factors such as the patient disease subtype.  For the sake of this tutorial, we will focus on the disease diagnosis, specifically people who are diagnosed with Crohn's disease (CD), people who are diagnosed with Ulcerative Colitis (UC), and people who don't have IBD are serve in the control group.  The patients in the control group are indicated by a (no) label.\n",
    "\n",
    "In this particular study there was exactly 1 rectal swab collected from each patient.  The number of people in each group are given below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CD    161\n",
       "no    107\n",
       "UC     31\n",
       "Name: diagnosis, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "md['diagnosis'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have been introduced to our data, we use a program in Stan to perform a Negative Binomial regression.\n",
    "The goal here is two fold: (1) determine if there is a difference between microbial profiles between the control group and the IBD patients and (2) determine which microbes can explain the differences between these groups.\n",
    "\n",
    "Briefly our proposed model given as follows\n",
    "\n",
    "$$\n",
    "y_{ij} \\sim \\mathrm{NB}(\\delta_i  \\lambda_{ij},  \\alpha) \\in \\mathbb{N} \\\\\n",
    "\\lambda_{i} = alr^{-1}(x_i \\cdot \\beta) \\in \\mathbb{R}^{d}_+\\\\\n",
    "\\frac{1}{\\alpha} \\sim \\mathrm{Gamma}(1, 1) \\in \\mathbb{R}_+ \\\\\n",
    "\\beta \\sim \\mathcal{N}(0, 10) \\in \\mathbb{R}^{p \\times d - 1}\n",
    "$$\n",
    "\n",
    "where $y_{ij}$ corresponds to the microbial counts for taxa $j$ and sample $i$, and $x_{i}\\in\\mathbb{R}^{p}$ corresponds to the covariates within each sample. $\\delta_i$ corresponds to the sequencing depth (i.e. number of reads sequenced for each sample), $p$ are the number of covariates, $d$ is the number of microbes, $n$ is the number of samples.  $\\lambda_{i} \\in \\mathbb{R}^{d}$ corresponds to the rate parameter of the Negative Binomial distribution, which here translates to the predicted proportion of microbe $j$ in sample $i$.  $\\alpha\\in\\mathbb{R}_+$ is an overdispersion parameter that can help better model the variance of the data.\n",
    "\n",
    "The quantity we are most interested in is $\\beta$, since this will give us an estimate of the __log-fold change__ in this case study.  Here we have placed a weakly informed normal prior over $\\beta$ in the hopes that it will be able to handle the small dataset since given here.  One thing that is important to note here is that we are using the inverse ALR transform instead of the ALR transform; this is intentional since our dataset is very sparse (mostly zero counts), we need to be able to handle these zeros.  Rather than trying to impute the zeros, we will utilize use Stan to explicitly model the generative process.  Doing so allows us to model the microbial proportions with the inverse ALR transform, which is then used to generate counts with the Negative Binomial distribution.  The take away here is that $\\beta$ is represented by the log-ratios represented in the ALR transform introduced earlier; we will later show how log-fold change can be extracted from the model estimate shortly.\n",
    "\n",
    "The Stan program behind this model is found below. We will then use Automatic Differentiation Variational Inference (ADVI) to fit this model. If you are curious about the details behind the GLM, feel free to check out the following blog post on this\n",
    "\n",
    "https://mortonjt.github.io/probable-bug-bytes/probable-bug-bytes/differential-abundance/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data {\n",
      "  int<lower=0> n;     // number of samples\n",
      "  int<lower=0> d;     // number of samples\n",
      "  int<lower=0> p;     // number of covariates\n",
      "  real depth[n];      // sequencing depths of microbes (in log counts)\n",
      "  matrix[n, p] x;     // covariate matrix\n",
      "  int y[n, d];        // observed microbe abundances\n",
      "}\n",
      "\n",
      "parameters {\n",
      "  // parameters required for linear regression on the species means\n",
      "  matrix[p, d-1] beta;  // covariates\n",
      "  real<lower=0> disp; // over-dispersion to help control for variance\n",
      "}\n",
      "\n",
      "transformed parameters {\n",
      "  matrix[n, d] lam;\n",
      "  // compute log(alr^{-1}(x * \\beta))\n",
      "  lam = append_col(to_vector(rep_array(0, n)), x * beta);\n",
      "}\n",
      "\n",
      "model {\n",
      "  // setting priors ...\n",
      "  disp ~ inv_gamma(1., 1.);\n",
      "  to_vector(beta) ~ normal(0., 10.); // uninformed prior\n",
      "  // generating counts\n",
      "  for (i in 1:n){\n",
      "    for (j in 1:d){\n",
      "      target += neg_binomial_2_log_lpmf(y[i, j] | depth[n] + lam[i, j], disp);\n",
      "    }\n",
      "  }\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(open('nb_diff.stan').read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In our particular use case, we only have 3 covariates, namely the 3 diagnostic groups UC, CD and control (no).\n",
    "As a result, the main thing we want to do is to be able to estimate the log-fold change between the groups.\n",
    "We can use statistical libraries such as patsy to design a constrast matrix, where we will represent these differences as one hot encodings.  This covariate matrix will then be represented by 3 variables, namely `Intercept`, `CD vs Control` and `UC vs Control`. This can be done with the code below.\n",
    "\n",
    "A more detailed discussion behind design matrices can be found at \n",
    "\n",
    "https://mortonjt.github.io/probable-bug-bytes/probable-bug-bytes/regression-intercept/\n",
    "\n",
    "http://mortonjt.blogspot.com/2018/05/encoding-design-matrices-in-patsy.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Intercept</th>\n",
       "      <th>C(diagnosis, Treatment('no'))[T.CD]</th>\n",
       "      <th>C(diagnosis, Treatment('no'))[T.UC]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1939.MGH100695.a</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939.MGH100695.b</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939.MGH101446</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939.MGH101666</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939.MGH102261</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Intercept  C(diagnosis, Treatment('no'))[T.CD]  \\\n",
       "sample_id                                                          \n",
       "1939.MGH100695.a        1.0                                  1.0   \n",
       "1939.MGH100695.b        1.0                                  1.0   \n",
       "1939.MGH101446          1.0                                  1.0   \n",
       "1939.MGH101666          1.0                                  0.0   \n",
       "1939.MGH102261          1.0                                  0.0   \n",
       "\n",
       "                  C(diagnosis, Treatment('no'))[T.UC]  \n",
       "sample_id                                              \n",
       "1939.MGH100695.a                                  0.0  \n",
       "1939.MGH100695.b                                  0.0  \n",
       "1939.MGH101446                                    0.0  \n",
       "1939.MGH101666                                    0.0  \n",
       "1939.MGH102261                                    1.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from patsy import dmatrix\n",
    "disease_type = dmatrix(\"C(diagnosis, Treatment('no'))\", md, return_type='dataframe')\n",
    "disease_type.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have introduced the model and the data, we will coerce the input data into a format that is recognizable by Stan.  CMDStan takes in as input a json file that is written to disk, which can be achieved by the following code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "nb_data = {\n",
    "    'n' : len(table),                                     # number of samples\n",
    "    'd' : table.shape[1],                                 # number of microbes\n",
    "    'p' : disease_type.shape[1],                          # number of covariates\n",
    "    'depth' : np.log(table.sum(axis=1)).values.tolist(),  # log of sequencing depth\n",
    "    'x' : disease_type.values.tolist(),                   # covariates represented as one-hot encodings\n",
    "    'y' : table.values.astype(np.int64).tolist()          # microbial counts\n",
    "}\n",
    "with open('data.json', 'w') as f:\n",
    "    json.dump(nb_data, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have our data formatted, now we can run ADVI and try to obtain model estimates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:cmdstanpy:found newer exe file, not recompiling\n",
      "INFO:cmdstanpy:compiled model file: /home/juermieboop/Documents/interviews/bch/short-introduction-to-ALR/ipynb/nb_diff\n",
      "INFO:cmdstanpy:start chain 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: TBB_CXX_TYPE=gcc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:cmdstanpy:finish chain 1\n"
     ]
    }
   ],
   "source": [
    "%set_env TBB_CXX_TYPE gcc\n",
    "import os\n",
    "nb_model = CmdStanModel(stan_file=os.path.join(os.getcwd(), 'nb_diff.stan'))\n",
    "fit = nb_model.variational(data='data.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once we have obtained model estimates, we can also sample from the posterior distribution.  \n",
    "This allows us to quantify the uncertainty surrounding our model parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from util import extract_log_fold_change\n",
    "beta_samples = extract_log_fold_change(fit, n=table.shape[1], p=disease_type.shape[1])\n",
    "# Compute means and std dev surrounding our log-fold change estimates\n",
    "beta_mean = beta_samples.mean(axis=0)\n",
    "beta_std = beta_samples.std(axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To evaluate which column of beta corresponds to the log-fold change between CD compared to C.ontrol, \n",
    "we first need to remind ourselves how the design matrix was formatted.  Looking at the columns of the design matrix,\n",
    "we can see that the second column corresponds to the CD vs Control covariate.  So we will pull out that\n",
    "column of the beta variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Intercept',\n",
       " \"C(diagnosis, Treatment('no'))[T.CD]\",\n",
       " \"C(diagnosis, Treatment('no'))[T.UC]\"]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_type.columns.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By default, we chose the first feature to be the reference feature.  If we want to match \n",
    "the log-fold change values with the microbial names, we can designate the first feature to \n",
    "have a zero log-fold change as mentioned in the previous notebook.\n",
    "\n",
    "We will now tidy up the CD vs Control log-fold change values and visualize the means and standard deviations for the log-fold change estimates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, '$log(Control/CD) + K$')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAD5CAYAAAAndkJ4AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAm2klEQVR4nO3deXTddZ3/8ef7LtnTJm3ThS4UKC1rKVJ2lEFxjjo6Ki4dBhVhsMC4os78dBwBHRTEEdRRRlCRVUFEXEBF9kE2aeleaCnFtumWtGnSbHd///64NyWEpLm3uTf3Jnk9zsnpvd/7Xd7JOf2872c3d0dERCRQ7ABERKQ0KCGIiAighCAiIhlKCCIiAighiIhIRqjYAQzFpEmTfPbs2cUOQ0RkRFm6dOkud2/oe3xEJ4TZs2ezZMmSYochIjKimNmm/o6ryUhERAAlBBERyVBCEBERQAlBREQylBBERARQQhARkQwlBBERAZQQREQkQwlBRGSEWXTjMyy68Zm831cJQUREACUEERHJKJmEYGYVZvZXM1thZmvM7GvFjklEZCwppcXtosBb3b3DzMLAX8zsj+7+bLEDExEZC0omIbi7Ax2Zt+HMjxcvIhGRsaVkmowAzCxoZsuBJuAhd3+un3MWm9kSM1vS3Nw87DGKiBTTjrYIa7a10R6J5/3eJZUQ3D3p7guAGcBJZnZMP+fc5O4L3X1hQ8Mb9ncQERnVHlvXREc0STBgeb93yTQZ9eburWb2OPAOYHWRwxERKbqeeQfjK8OUBQNUhoN5f0bJ1BDMrMHM6jKvK4GzgZeKGpSISAlJufPUhl3UVYUxG901hGnArWYWJJ2ofunu9xc5JhGRktEeSdAZSzK9vrIg9y+ZhODuK4Hjix2HiEipau2KURYKMK4iXJD7l0yTkYiI7F9rd5xTDp1YkA5lUEIQERkRIvEkkXiKs+YVbnRlyTQZiYjIwFq70/MOzpo3mQtOP6Qgz1ANQURkBGjtilMRCjB7UnXBnqEagohICemZb3D3xafue339ogXs7Y4zZVxFQZ+tGoKISIm74fENAEwdX17Q56iGICJSAgbaAS2aSPLL5xtpqC2nPJT/2cm9qYYgIlLCtrVGcJyDxhe2uQiUEERESlY0kaS5PcoHT5hJeQHWLupLCUFEpEgW3fjMG5qKWrtirNraxhnfepRVW/cC8MmzDhuWeJQQRESGUX9JoEdbd5z1TR2k3Dlp9gQmVpcxZ3INM+qrhiU2dSqLiAyDgZJAj73dcdbvbKcyHOSIqbVct2jBoNfkm2oIIiJFtnprG+t2tlMeSieDcLA4RbNqCCIiRRRNJLnwlucJBQL7TQZ3X3xqwWNRQhARKZDes477k0g563d2EDRj3tQaykLFbbRRk5GIyDBxd9q642xp6eLSO5ayemsbXbEkPzzvTVSVFf/7efEjEBEZ5SLxJFtautjVESOWTGFARTjdgTxrQhVvmdvADx9LL0/RuzYxHM1EvamGICJSYNc/tJ5tbRGqyoLMaajmhIPrefSLf8e8qbVMqC4rdnj7qIYgIlJAHdEEP35yIw215RxawKWr80EJQUSkQFLubGzuZHJtBdPr978W0XA3D/VHTUYiIgWyrTVCdzzJVe87hlCg9Ivb0o9QRGQEiiVSbG/rZkJ1GWcfNaXY4WSlZBKCmc00s8fM7EUzW2Nmny12TCIiB2rdjnZSTkl1Gg+mlPoQEsAX3P0FM6sFlprZQ+6+ttiBiYjkavmWPQDUlL1+2epiDisdTMkkBHffDmzPvG43sxeB6YASgoiMOMu3tBEK2L7Zx6VW+PenZJqMejOz2cDxwHP9fLbYzJaY2ZLm5uZhj01EJBsrGlupKQ9hZsUOJWsllxDMrAa4F/icu+/t+7m73+TuC919YUNDw/AHKCIyiL2ROK80d1BTXjKNMFkpqYRgZmHSyeBOd/91seMRETkQqxrbcIdqJYQDY+l61U+BF939umLHIyJyoJZvaQWgprzw+yDnUymlr9OBjwKrzGx55th/uPsfiheSiEhuFt34DOt3tnPIpGru/dfTix1OTkomIbj7X4CR0/siIpLRd9+DjmiCM+eOvD7OkmkyEhEZDaKJFPGkc9zMumKHkrOSqSGIiIw0PTWD3jqjCQAWjMCEoBqCiEgedUQTGHDktHHFDiVnSggiInng7vzsqVfZ3hahtiJERXhkjTACNRmJiAxZKuW8uruTr/1+LXVVYQ5rqCl2SAdENQQRkSFwd17Z1cmujhiXnT2XuZNrCAVG5oBJJQQRkX4suvGZfZ3GvV/3ta0tQktnjJn1lXz27MNH1NpFfSkhiIjkoHdyaOmM0binm4nVZUwbv/8tMkcC9SGIiByAzbu7eKW5g+qyIIdOqt5XMxgJy1wPZNCEYGZBd08ORzAiIsU2UNNQb+7OV36zCoDDp9QQGKF9Bn1l02T0k4JHISJSRPvrI+jP7s4YT768i5kTqigPjbzhpQPJJiEEzezrfQ+aWa2Z3VOAmERECibXwr+veDLFpt1dHD+rjim15XmMrPiySQgXAqea2UU9B8zsOGAJ8GKhAhMRKaTBEkMimWJbazfffvAlNrd0sWl3J1+8ZwUv7WgnmXKuPufY1/UbjOS+gx6D9iG4e8LMzgEeN7NtpPc5/gpwkbs/XOgARUSGWyyRYt3OdrpiSW58YiPJlGMGT2/YBcDsiVUcMXXkLU0xmEFrCGZ2A3AecC3wM2ARcLKSgYiMRpF4krXb9xKJJzliai0bvvkuTjpkAifOnsDTX34bx04fz+RxI3+IaX+yGXa6HJgPHAuEgbnAj81sFbDK3e8qXHgiIkPXd7+Cvprao+zuiPLmax9lS0s3oYBx5LRxI25P5KHKpsnopt7vzWwGryWIfwCUEERkRHJ3/uv+tby6q5PKcJDjZ9aTTDoNteUjcnG6oco5/bl7I9AIaGtLERmxEskUG3d18te/7WHquHJmTaji++cen9MIpNHQkdzb2KoPicio1rtpaKBmotauGM0dMfZ0xXCHr777KP68Zsewx1qKck4IZvZ2d3+oEMGIiByIbL/V3/yXV1m3s4NQwJhcW05DTTn/csYhOSWE0VYr6O1AagjfApQQRGTYZVMDGMjujihfv38t9VVh5kyuITDIqqSjueAfiJqMRKTkDWVmMUBbd5xXmjs56ZAJeMoHTQZ9jZXkkFVCMLOfAQ4YMMvMbu75zN0vzFcwmfu+G2hy92PydV8RGZvcnVueepV1O9qpCAf58ccWsvi2Jfu9ZqwU/v3JtoZwS6/XZwC35j+Ufc/5AXBbge4vIiUs12aggbg70XiSTS1dXPn7tdRVhjm0oZrxleF+zx/LSaC3rBKCuz/R89rM2nu/zyd3/z8zm12Ie4vI6JJypzuW5L5ljWxp6SKZclJAMpnitGseZXtbBAP+8x+O5M9rdozoncyGy4H0IcTyHkUOzGwxsBhg1qxZxQxFRIZgsNrAQJ+7O79bsY0VW9qIJVNcdvcKDAgGDDMImHH64Q2sbGylrjLMRW8+lIfW7nzD/VUreKOct9B091MKEUgOz7/J3Re6+8KGhoZihiIieZDNctTdsSS7O2Nsb+tm7fZ2PvOLZYSCxpyGah7+/FtYOLueEw6u502z6lkws47/Ofd4po6rGJOzjYdCo4xEpKQ9uGYHK7e27XsfDhrf+sCx3Lu0ETNjzuTanEcNSf+UEERk2OTaaRxNJPn3X62kuizIIZOqKQ8FCAUDLDpxFr9+YesBxaCmooHl3GRkZtVmVpB6mJn9AngGmGdmjWb2L4V4joiUPnfnlaZOEskUcybXUF0eIhTMuciSHAxaQzCzAPBPpPdEOBGIAuVm1kx6gbub3P3lfATj7ufm4z4iMnIlkimeeWU3r+7qoj2a4LuLFvCLv24e0j1VK8hONk1GjwEPA18GVrt7CsDMJgBnAdeY2X3ufkfhwhSRkWqgZqKUO8mUk0g60USSq+5fy8rGNrrjSc798bMATK4t533HT+83IQxWyCsJ5C6bhHC2u8f7HnT3FuBe4F4z63+2h4iMWf2NHHJ3OqIJvnjPCpZu2kPKX/vs1V1dVIQDTKyp5BvvP5bvPbyecJZNRCr882PQv3ZPMjCzGWZ2rJlVD3SOiIw9vYeN7m8IaSrlrG/qYO32dv60egcTq8uZPbGKOQ3VHDG1luVXvJ0jp41jel0lZ85tyDoZSP5k04cwG/g1MAWIAFPN7FHgC+6+vrDhichocfuzm2jtijO9roKHPn8mF/zs+dd9XlU2eIOFagKFlU2T0beAG939RgAzCwEXAn80s3fkq0NZREaOXIePdseSXP3HFxlfGWZ6XWVWhX9vSgTDI5s62dyeZADg7onMPsuXApcXLDIRGRXcnVeaO6gIBzl0UrXWFCph2SQE7/eg+5+BI/MbjoiMZMmU0xVL8MT6ZnbujbChqYPlW9rojCW56n3HUBZSv0Apy6beNjUzQWw1sMbdO3p91m+yEJHRaaAO41gixdV/eJEXNqdHDp1/81+B9DITtRVh6qvCvHv+Qdz+zKasn6VmouGXTUK4Ejge+BhwjJntBdaQThBTCxeaiJSCwfoLdrRF2NzSxYrGVuqryqivCnPNB+Zz1QNrKQsGBm0iUsFfOgZNCJn+gn3MbAYwHzgWKMi+CCJSHH1rAIMV1v+3vplNLV3UVYb53afO4N9+tQKAhbMnUB564wo3mkxW2rIZdvpVoMvdvwPg7o1AI+llK0SkBOWy18CB7lIWjSf5zF3LqAwHmTO5hlkTq3K6XoV/6cmmyeijwIK+B83sIqDB3a/Od1Aikj/52pYSoCuWIJpI4Q7bWrsJBIy5U2oIBjRyaDTIpsu/2927+jl+O/CRPMcjIiXq7uc3s2rrXtbv7ODlpg46Y0mu+/ACbUIzimRTQ+g2s2nuvr33QXePmlmiQHGJSJ4NpabQ3B7lS79exfjKMDPqKzEgHAzw9qOm8JMnN+73WjUNjRzZJITvAL81sw+5+74xY2Y2GUgVLDIRGVB/fQA98lkAR+JJtrd1s7mlmzcfPolILElgP81DvZ+tRDDyZDPK6B4zqwKWmtmzwHLSTU0fIj0kVURGEXcnEk+xuzPGGd96lF0dMcZXhvnxxxbum18go1M2o4xOBW4jvcDd+4GjgU7gXHdfUtjwRKRHPjuHe4slUnRGE1z30HrW7WinI5ogkVmX+q1HTGZbazfjKkID9hWoJjB6ZNNkdD7wQ2A98CfgenffUdCoRAQYehKIxJO0dMYyBbzjDj98bAObW7qIxJOc/M2H2bk3CsCGR1+mPBSkvqqMmooQ4ypC3PzxEwecnSyjTzZNRpcAmNkRwDuBW8xsPOmd1P4EPOXuyYJGKTKC5FKIDzQfYKiFcDSepLkjyunXPMruztjrPvv2g+sIGJSFAvzdvMm8sGkP1eVBfnXpaW9Ykrqvnt9JSWJ0ynoNWnd/CXgJuN7MKklvn/kh4DpgYWHCExk98lng9xVLpGjrjrM3Euf0ax5la2s3AGcfOYXGPemdyAAM4xeLT9nXF3D9ogX7Ysl1SWoZfbLpQ7iNdEfycmCFu+92927SM5U1W1nGnEJ9q89Gyp37V27jpR3txJMp3vm9J9nY3EE0kR7wFwoYbz9qAqGAUVcV5ifnL3xDXJo3IAPJ5ivBrcBxpPsSjjOzWuBFXksQ9xQuPBGB9MifHW0RtrV18/zPl1EeClBVFmRGfSU727qZMq6ccRVhqsqC/O9HTih4clJH8uiUzZ7Kj7j7de5+vrsvAOYB/wGsA07OZzBm9g4zW2dmG8zsS/m8t8j+DLQX8P72CB4u8WSKT9y2hE0tXVSGg/zs4ydy3IzxzJ1Sy48/tpDDp9QybXwl1eUhbT4jQ5J1o6GZTQAuAyYDa4Hb3P32fAViZkHSo5neTnrxvOfN7HfuvjZfzxDJRn+dwoUa8tlbz/j/3yzbyuaWLhIpJ5Vy9kbiGMbBE6qYMq6cs46YzI+eeKVgcYAmmI1VufQi3QU8DDxHeunrv5jZBe6er5kqJwEb3H0jgJndBbyXdPIRGXHiyRSbdnfy6q4utrV2s7mli2TK+fzdy3l5ZzspwD29y9hpVz/CtrYIAJ+7ezkGhIJGwIyKUJCff+IUrnpgaP8VVLDLYHJJCNPc/drM6/vN7G7g58ApeYplOrCl1/tG8twkJdJXPpuD3J2OaIKWzvRon6Mvf5BY8rXVXQwIBoy//q2FrniSgBlmEDTjlMMm8uwru6kIp/sAvvqbVa9r/jl2xvi8xbk/ShpjWy4JocXM5rv7SgB335hZ0iJf+mv8fMMWnWa2GFgMMGvWrDw+XsaKfDX/vLyznY3NHXTFkpz134+zeXcXSXcMqK0I8dFTD2belFoObahhel0ln7xzKWbW74ik6z782vDPeVNr89oXoE1pJFu5JITFwL1m9iSwCjgKyGdDZiMws9f7GcC2vidldnC7CWDhwoXa01mGVc94/517I7z9+v8jYOnC/5jp4+mMJhhXEWJ8VRmhgPHldx75umsLXcjnUrArCUh/skoIZhYAPgC8CXgf6WSwAvhiHmN5HjjczA4BtgL/BPxzHu8vo0ShOngTyRRNeyNEE0mSKeexl5po2hshnnKSKSeaSHHCVQ/RHkkQDBifeesc/rJhF+FggP859/iiDfVU4S75klVCcPeUmZ3t7t8EflmIQNw9YWafAh4EgsDN7r6mEM+S0Wew5aD3l0Qi8SSbdnfR2h3npG8+su/4Bbe8toyDGYQDAd53/EEs39LK+Iown//7eTz3akshfp0hU5KQA5HNjmk9lpnZFVbAgc7u/gd3n+vuh7n7Nwr1HBkdhjpHoLk9SuOeLlZubWNvJM70ugq+8f5jOGRSFXMaqrn30tNYMHM8Jx5cz0mzJ3D8rDqu/eBx1FeV7XdPgHy4++JTVajLsMulD2Em6eGml5rZc8BKYKVmKstwyGczUWtXjJ3tUU65+hGSKWdidRmzJlRRFgpw3skH87vl6a6rEw6upzyUv2Ue+hvb3zuhDbVfQGSockkI/+bum8ysnPSeCMeSnjughCAjQjyZ4rN3LWPdzg7CQeMTbz6UpzfsorKseGv7qMCXUpJLk9F9kN5L2d1fcPdbgXsLE5ZI/paNiMSTNLVHWbW1jQdWbmd6XSULZtbxpXceUdRkIFJqslnt9MOkRxfVmtmRwPpe+x/cBMwvYHwiOXN3EimnO5ZkbyS+bxvI6rIg91xyGl/7ff7GKmiJBxlNsmkyegqoAC4ivffBPDNrJT1HoLtwoYlkJ5VyIvEk7ZEEn71rGcu2tBJPvjZFpfc2kEcdNO6AnqHCXsaCbHZM2wrcZmavuPtTsG+hu0NIb5gjMmya26NsbukilkiRTDnxZIqjr3iQ7ni60ro3EmdcRZia8hCVZUEqw0FtAymSpVw6lcvN7KdAK7Ca9CijRCGCkrGnv7kDfe3uiPL31z9Ba1ecslCAYMAIB40PvGkmj61rorosyO8+dQbn/vjZnJ8/UNOPRv7IWJJLQrgD+GTmmvmkZywfDczJf1gyFmQ7lHRrazfrd7azpyvOcTPGM6Meqnp1Bl/+nqNYs60NIOf5ASrcRV6TS0LY4O73ZV5rqKkUXCrlbGuL8LbvPE4skWJGfSX3Xnoa5/3kuSHdV0lApH+5JIQnzOwy4LvurkXlZNBv+EPZaKapPcKa7XvpiiV59/xpNO7pojwUJBTMZaT0a5QERAaXS0I4GjgG+H9mtpT0nsrLNVNZehuowM+lU7c7luScG54mEk8yd0oNP/jnN6lTWGQYZJ0Q3P0cADOr5LXkcApqPhpTCrW9pLvTHk3Q1hWnqT1KXVWYI6eNo6Y8l+8sr6dagUhuhrqn8i0FiktKSKGWm06mnNbuOJfdvZylm1tJptItkeMqQtx76Wn8+69WHtB9lQhEDkwp7aksRdLf0tH9LSOdT0++3MyKxvQEst0dUeqrwtRVlTG+MkQoEODgidWD3kOzhEXyK5ceumnufq273+/uVwPvAb5foLikwPK1TlAu3J1oIsnfdnXy0Z/+lWDAOGJqLc9/5WwOa6hhYnUZocCBdRqLyNCV0p7KMgwKnQTcHQc6owniyRTdsSQ/fGwD63a00xFNkMg0C114+iGsamwlELCcRw6pNiBSGLkkhIuBXxVwT2UpccmUk0imiCRS3P7sJjbu6iSRTC8hkUw5p1/zKNvbukk5HH3Fg/uue3HHOirCAeqqwlSXhxhXEeLy9xyVU3JSEhApvFxGGb1kZoXcU1lKSFt3nMY93XTFEkTiKaKJJIf9xx/2ff7V36wmFDDCwfQSEqFggJMPmcDTr+wmFDA+eurB3PncZirCAe5efCqX3LG0iL+NiGQjm+Wvvwp0uft33D1GgfZUlvwaysig9kicd33vSba2dlMRDlARDjK+MsSiE2dx37JGykJBbrngRD5/93J676h63aIF+5578ZmH8ehLTQDUV5dl/Wx1FIsUTzY1hI8CC/oeNLOLgIZMB7OMUD1LR0fiSe58bhOv7uqkqT3KrAlVHH3Q6+cBXPb2uTy7cTcAM+qryOf22ir8RYovm4TQ7e5d/Ry/HXgBUEIoIdm2y7dH4mxu6eaYKx+kK5ZeOvor960GYFJNGQ985gwuunVJweIUkdKTVUIws2nuvr33QXePmpmWvx4B+jYfdUQTvLSjnVAwwHknz+LxdU1UhIP89PwT+dxdyzAzaivCBY9LtQKR0pJNQvgO8Fsz+5C7b+o5aGaTgVQ+gjCzDwFXAkcCJ7m7vprmIJf+gs5MMggHAxw1bRxXvOdo1m7bC8BBdZV5bQbqTYW/SOnLZse0ezLzDZaa2bOkF7ULAD2FeD6sBs4BbszT/aQfj69r4sUd7QQDxpHTaikL5WcSmDqCRUaHrIaduvutZvZr4P2kF7brBM7N1zd5d38RKNi307HO3Wnc083Hf/Y8leEgc6fUUB4KDnrdUAt3JQeRkSWbYafmae3Abfs7J+/RyYAGayZKpZy2SJz/un8tK7e2EYmnWLRwJhubO3LeVaw31QZERq9s2gweM7NPm9ms3gfNrMzM3mpmtwLnD3YTM3vYzFb38/PeXAI2s8VmtsTMljQ3N+dy6ZgQS6TYsqeLZVtaWb+zgzue3URZMMBhDdV864Pzh5QMRGR0y6bJ6B3AhcAvzOwQoBWoJJ1M/gxc7+7LB7uJu5994GG+7j43ATcBLFy4cEzUSrLZgB7g6Q27WN7YijvUV4WZXFvObz91BuffnP2CtPrWLzJ2ZdOpHAFuAG4wszAwifTchNYCxzam5TrTOJpI8elfLKM8FGDelFoqwuk+gp5/+1LTj4j0ldMwE3ePu/v2fCcDM3u/mTUCpwIPmNmDg10jr0m5s6GpPb3l5OTaAZOAiMj+5LJj2sukVzldQXro6Qp3/1s+gnD3+4D78nGvkWagzWmyFU+m+NuuLjqiSW44703c+vTfChGmiIwBudQQbgR2ALuBdwKrzWyVmX0905Qkw6ytO847v/ckzR1Rpo2v4F3HTuv3vLsvPlXNQiIyqFz2Q/iIuy/oeWNmPwIuAPYC1wGfzm9o0lvKndauOK3dcc6/+a+sbGyjO55k5oRK5k6pob4q+xVFRUT6k0tCaOuzY9pyMzvF3T9rZi8UKL4xr6k9wqbdXezqiJJIOcGA0dIZozwcYGJNGQ9ddqZGEYlIXuSSEC4B7jCz5aT7EObx2lpG+nqao8FWJU0kU2xvi3DmtY8TiSepqwrTUFtOXWWYX15y2r7r1YEsIvmSy45pL5rZSaTXHJoPbACuMLNq4K4CxTfquTtdsSRPrG+mpTNGJJ7k3Jue5YXNrTjwj8cdxKbdnVkX/KoBiMiBymWU0QTgMmAysBa4zd33ZD6+qgCxjTr9jSLa2trN1tbI65p9qstDTB1fwcTqMr5/7vE57T0sInKgchlldBfQDvweqAL+kqkxyH4suvGZAQv0HW0RtrZGmFRTxr2XnsoxB43j+Jl1/Olzb2HWhCqqy3Np0RMRGZpcSpxp7n5t5vX9ZnY38HPglPyHNfrt6oiyqaWL+qowh06q5oSDJygBiEhR5VICtfQZZbQxs0+C9GN/zTxPvtzMxuZOaitCzGmo0bLfIlISckkIi4F7zexJ0jOWjwZeKUhUo1hnNMElty+lsiy9L0Euq48O1GGsjmQRyYds9kO4jfQw0+XAW4G/I73V5TLgC4ULbfSJxJOs29nO5NoKJteWEwpk14XTX4GvJCAi+ZZNDeFW4DjSex4cB4wjPcooDLwHuKdg0Y0S0USSK3+3hlVb2zAzbrngRP7zN6vfcJ4KeREppmyWv34EeKTnvZmFgKNIJ4eTUUIYkLuzvS1C455uVm/dy4TqMqbXVXL4lNp956gZSERKRc67rLt7wt1Xuvvt7v7FQgQ1GqTceXV3F1v2dFNXFeaJfz+LwxpqNLNYREqWxjnmUc/IojsuOpn1O9tp605w0PgKZtRXMr2u8nXnqgYgIqVGCaEAHlyzg7buBAdPrGLquIpihyMikpWcm4xkcLc/s4nyUIApteXFDkVEJGuqIeRZVyzBqq17mVlfqQlnIjKiKCEMUd8F65r2RikLBWjop3agfgMRKWVqMjoAAy1Yl0w5zR1R3n3sNMJB/WlFZGRRDSGPdnVESTmcd8rBXPunlwDVCkRk5FBCyMH+Fqxr6YyxvS1CVVmQN82qG76gRETyRAkhD5Ip54JbnieWTHFEQy1mppqBiIw4JZEQzOzbpNdFipFeQfUCd28talCDcHe6Y0niyRQbmjrojCU5vKGGcRXhYocmInJASqXn8yHgGHefD6wHvlzkeParpTPGkk17OPLyP/HC5lb2RhJcc86x1FeXFTs0EZEDVhIJwd3/7O6JzNtngRnFjGd/OqMJXmnuoCIc5EvvPIJZE6o4YmotH1o4s9ihiYgMSUkkhD4uBP440IdmttjMlpjZkubm5oIH03uIaTyZYv3ODoIBY96UWi458zCmja9gfKWaiURk5Bu2hGBmD5vZ6n5+3tvrnK8ACeDOge7j7je5+0J3X9jQ0DAcofc8l5ebOoinUsydUktZqBRzqYjIgRu2TmV3P3t/n5vZ+cC7gbe5uw9PVNl7ZuNu2iMJZk+soqa8JPriRUTyqiRKNjN7B/D/gDPdvavY8fTnzmc3EwwYDTUDL1inoaYiMpKVSrvHD4Ba4CEzW25mPypmMH2XpoglUjy4ZgcNNeUEAlqwTkRGp5KoIbj7nGLHsD/N7VESKWfKuHTtoHdNQLUCERktSiIhlDJ3p6k9ypsPn0QskSp2OCIiBVMqTUYla09XnFgyxXknH1zsUERECkoJYT/W72xnc0sX4aBx9pGTix2OiEhBqcmol94dybs6orz3B0+RcufwyTWEggH1F4jIqKaE0I/m9igbd3Vy0uwJxJMpTUITkTFhzJd0fYeYujtbW7upLgty5ydOVjIQkTFDpV0fuztjRBMpDqqr1DaYIjKmqMTrxd3Z3hahIhygvkoL1onI2KKE0Etrd5yuWJKDxldiphnJIjK2KCFkuDvbWiOUBQNMrNFGNyIy9ighAIlkik//Yhkd0QTT6ioIqHYgImPQmB922h6Js6GpkxWNbcyor2RK7etXM9XcAxEZK8Z0QkimfN8OaL++9DS+8cCLgJKAiIxNY7rJaGVjK4mUM6O+kvkz6oodjohIUY3JGkLPRLSTD50IQJ32RBYRGds1hMfXNVFTHiKkCWgiImM3IcQSKVY2tlGnCWgiIsAYTght3XFAzUUiIj3GZB8CpGclT64tp6osuO+YRheJyFg2JmsIKXfauuKcNW+ylqgQEckYkzWEjkiCpDtnHTGZdxwztdjhiIiUhDFZQ2jtjmPA6XMmFjsUEZGSURI1BDP7L+C9QApoAj7u7tsK9bzaihBBq6S2Qh3KIiI9SqWG8G13n+/uC4D7gcsL+bD6qjKm11cW8hEiIiNOSSQEd9/b62014MWKRURkrCqJJiMAM/sG8DGgDThrP+ctBhYDzJo1a3iCExEZA4athmBmD5vZ6n5+3gvg7l9x95nAncCnBrqPu9/k7gvdfWFDQ8NwhS8iMuoNWw3B3c/O8tSfAw8AVxQwHBER6aMk+hDM7PBeb/8ReKlYsYiIjFWl0odwjZnNIz3sdBNwSZHjEREZc0oiIbj7B4bzeVqzSETkjUqiyUhERIpPCUFERAAlBBERyVBCEBERQAlBREQylBBERARQQhARkQwlBBERAZQQREQkw9xH7tYDZtZMeqkLERHJ3sHu/oblokd0QhARkfxRk5GIiABKCCIikqGEICIigBKCyD5mdqWZrR6G57iZfbDQzxHJlRKCjGpmdkumAP5JP59dm/ns/syh/wbOHN4IRUqHEoKMBVuARWZW3XPAzELAR4HNPcfcvcPddx/oQ8ysbEhRihSZEoKMBSuBl4EP9zr2D0AEeLznQH9NRmZ2vpmtMrOome00s1t6feZm9kkz+7WZdQLfzBy/2Mw2mFks8+8n+olpqpk9YGZdZrbJzD7S57nTzewuM9uT+Xmg997jZjbTzH5rZi2Ze7xkZv90wH8hEZQQZOz4KXBhr/cXAj8DBpyIY2YXAzdmzpsPvAtY0+e0K4A/AMcCPzSz9wM/AL4LHAN8D7jBzN7T57qvAb8DFgA3AbeZ2cLMc6uAx0gnrDOBU4HtwMOZzwBuAKqAs4Cjgc8BrYP8DUT2SxPTZFTLfKOfRLp5aBvpgr2d9Az3w4GvA5Pc/d1mdiXwQXc/JnNtI3CHu39pgHs78AN3/3SvY08B69z9wl7HbgHmuPsZva77ibt/otc5DwM73P0jZnYh8GVgrmf+g5pZEGgCLnX3X5rZSuBed//aUP9GIj1UQ5Axwd33APeRrhmcDzzu7psHOt/MJgPTgUcGufWSPu+PBJ7qc+wvwFF9jj3Tz/uec04ADgHazazDzDqANqAeOCxzzveA/zSzZ8zsKjM7YZA4RQYVKnYAIsPoZuBWoAO4fJBzLct7dvZzrL9qdy5V8QCwHOivT6AFwN1/amYPkm7GOht42syudvcrc3iOyOuohiBjySNAjHQT0m/2d6K77wS2Am/L8RkvAmf0OXYGsLbPsVP6ef9i5vULwBxgl7tv6PPT0ivGRne/yd0/TDrBLc4xVpHXUQ1Bxgx3dzObT7rvLJrFJd8ArjezncADpDtx3+bu39nPNd8G7jGzpcCfgXcA5wHn9DnvHDN7nvQopw+STjwnZz67E/gi8Fszu5z00NiZwHuBH7n7y2b2PeCPwHpgXOY5fZOOSE6UEGRMcff2HM79XzOLAV8AvkW6ueYPg1zzGzP7NOkC/bukO6//1d1/3+fUK4EPAN8HmoEL3P35zD26zOwtwDXAPcB40h3ijwF7MtcHgP8hnSjaSdd+vpDt7ybSH40yEhERQH0IIiKSoYQgIiKAEoKIiGQoIYiICKCEICIiGUoIIiICKCGIiEiGEoKIiADw/wGZbYFv0wAv5gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "ctrl_cd = 1  # grab the log-fold change corresponding to CD vs Control\n",
    "differentials = pd.DataFrame({\n",
    "    'mean': [0] + beta_mean[:, ctrl_cd].tolist(), 'std': [0] + beta_std[:, ctrl_cd].tolist()},\n",
    "    index=table.columns)\n",
    "differentials = differentials.sort_values('mean')\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.errorbar(np.arange(len(differentials)), differentials['mean'], yerr=differentials['std'])\n",
    "ax.set_xticks([])\n",
    "ax.set_xlabel('Microbes', fontsize=14)\n",
    "ax.set_ylabel('$log(Control/CD) + K$')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On the y-axis of this plot, we denoted the log-fold change per microbe up to some unknown constant K.\n",
    "This constant is due to the change of the total biomass that we cannot directly measure.  \n",
    "\n",
    "It turns out that unless we are performing hypothesis tests on each microbial feature,\n",
    "it doesn't really matter which reference we chose.  The reason is because we don't necessarily know \n",
    "which microbe is changing in abundance.  As a result, our choice of reference won't necessarily be\n",
    "indicative of determining if a microbe has changed in abundance.\n",
    "\n",
    "Now, _if_ you are able to determine which confidence which microbe is not changing, then you will\n",
    "be able to make much stronger statements about microbial log-fold change.  But figuring this out is still an outstanding question.\n",
    "\n",
    "BUT, even though we can't figure out what K is, we can always sort the microbes according to their log-fold change, since ranking is a scale-invariant operation.  Recalling out notation in the previous section, where $A$ and $B$ refer to the ground truth abundances between two conditions and $p_A$ and $p_B$ are the corresponding proportions, the following statement holds.\n",
    "\n",
    "$$rank\\big(\\frac{A}{B}\\big) = rank\\big(\\frac{p_A}{p_B}\\big)$$\n",
    "\n",
    "This holds because the bias $K$ applies to every microbe equally.  As a result, we can still sort the microbes according to their log-fold change and identify the microbes that increased the most and microbes that decreased the most.\n",
    "\n",
    "So, let's who the top most increased microbes and top more decreased microbes are."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;f__Comamonadaceae;g__Alicycliphilus</th>\n",
       "      <td>-2.898736</td>\n",
       "      <td>0.261368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Lachnospiraceae;g__Epulopiscium</th>\n",
       "      <td>-2.597962</td>\n",
       "      <td>0.234430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Lactobacillus</th>\n",
       "      <td>-2.536195</td>\n",
       "      <td>0.244721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Veillonellaceae;g__Megamonas</th>\n",
       "      <td>-2.351506</td>\n",
       "      <td>0.217571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Adlercreutzia</th>\n",
       "      <td>-2.284217</td>\n",
       "      <td>0.284247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o__Bacteroidales;f__[Paraprevotellaceae];g__Paraprevotella</th>\n",
       "      <td>-2.071902</td>\n",
       "      <td>0.214061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Veillonellaceae;g__Megasphaera</th>\n",
       "      <td>-1.966108</td>\n",
       "      <td>0.193813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Veillonellaceae;g__</th>\n",
       "      <td>-1.964276</td>\n",
       "      <td>0.200099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Tenericutes;c__Mollicutes;o__RF39;f__;g__</th>\n",
       "      <td>-1.869300</td>\n",
       "      <td>0.233279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Bacilli;o__Turicibacterales;f__Turicibacteraceae;g__Turicibacter</th>\n",
       "      <td>-1.658010</td>\n",
       "      <td>0.232382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Lachnospiraceae;g__Anaerostipes</th>\n",
       "      <td>-1.583189</td>\n",
       "      <td>0.219043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Peptostreptococcaceae;g__</th>\n",
       "      <td>-1.542177</td>\n",
       "      <td>0.207183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Lachnospiraceae;g__Roseburia</th>\n",
       "      <td>-1.444132</td>\n",
       "      <td>0.186581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Lachnospiraceae;g__Lachnobacterium</th>\n",
       "      <td>-1.349103</td>\n",
       "      <td>0.214591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium</th>\n",
       "      <td>-1.345480</td>\n",
       "      <td>0.194259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Erysipelotrichi;o__Erysipelotrichales;f__Erysipelotrichaceae;g__Catenibacterium</th>\n",
       "      <td>-1.320741</td>\n",
       "      <td>0.244833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Moraxellaceae;g__Acinetobacter</th>\n",
       "      <td>-1.274457</td>\n",
       "      <td>0.231749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Clostridiaceae;g__</th>\n",
       "      <td>-1.241382</td>\n",
       "      <td>0.213407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__Desulfovibrionales;f__Desulfovibrionaceae;g__Desulfovibrio</th>\n",
       "      <td>-1.232206</td>\n",
       "      <td>0.266163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__</th>\n",
       "      <td>-1.179591</td>\n",
       "      <td>0.283026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        mean       std\n",
       "k__Bacteria;p__Proteobacteria;c__Betaproteobact... -2.898736  0.261368\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos... -2.597962  0.234430\n",
       "k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactoba... -2.536195  0.244721\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos... -2.351506  0.217571\n",
       "k__Bacteria;p__Actinobacteria;c__Coriobacteriia... -2.284217  0.284247\n",
       "k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o__... -2.071902  0.214061\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos... -1.966108  0.193813\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos... -1.964276  0.200099\n",
       "k__Bacteria;p__Tenericutes;c__Mollicutes;o__RF3... -1.869300  0.233279\n",
       "k__Bacteria;p__Firmicutes;c__Bacilli;o__Turicib... -1.658010  0.232382\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos... -1.583189  0.219043\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos... -1.542177  0.207183\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos... -1.444132  0.186581\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos... -1.349103  0.214591\n",
       "k__Bacteria;p__Actinobacteria;c__Actinobacteria... -1.345480  0.194259\n",
       "k__Bacteria;p__Firmicutes;c__Erysipelotrichi;o_... -1.320741  0.244833\n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac... -1.274457  0.231749\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos... -1.241382  0.213407\n",
       "k__Bacteria;p__Proteobacteria;c__Deltaproteobac... -1.232206  0.266163\n",
       "k__Bacteria;p__Actinobacteria;c__Coriobacteriia... -1.179591  0.283026"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "differentials.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Micrococcaceae;g__Rothia</th>\n",
       "      <td>1.092910</td>\n",
       "      <td>0.227672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Moraxellaceae;g__Moraxella</th>\n",
       "      <td>1.189515</td>\n",
       "      <td>0.259315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__[Tissierellaceae];g__Anaerococcus</th>\n",
       "      <td>1.221614</td>\n",
       "      <td>0.212661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pasteurellales;f__Pasteurellaceae;g__Aggregatibacter</th>\n",
       "      <td>1.297860</td>\n",
       "      <td>0.208737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Neisseriales;f__Neisseriaceae;g__Eikenella</th>\n",
       "      <td>1.309592</td>\n",
       "      <td>0.211117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacteriales;f__Enterobacteriaceae;g__Erwinia</th>\n",
       "      <td>1.389300</td>\n",
       "      <td>0.230333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Bacteroidetes;c__Flavobacteriia;o__Flavobacteriales;f__[Weeksellaceae];g__</th>\n",
       "      <td>1.390022</td>\n",
       "      <td>0.245553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;f__Burkholderiaceae;g__Lautropia</th>\n",
       "      <td>1.395832</td>\n",
       "      <td>0.190013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacteriales;f__Enterobacteriaceae;g__Serratia</th>\n",
       "      <td>1.533259</td>\n",
       "      <td>0.254360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pasteurellales;f__Pasteurellaceae;g__Actinobacillus</th>\n",
       "      <td>1.639519</td>\n",
       "      <td>0.219621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Neisseriales;f__Neisseriaceae;g__Conchiformibius</th>\n",
       "      <td>1.709216</td>\n",
       "      <td>0.159673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Bacteroidetes;c__Flavobacteriia;o__Flavobacteriales;f__Flavobacteriaceae;g__Capnocytophaga</th>\n",
       "      <td>2.121370</td>\n",
       "      <td>0.238828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o__Bacteroidales;f__Porphyromonadaceae;g__Porphyromonas</th>\n",
       "      <td>2.255435</td>\n",
       "      <td>0.208298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__[Tissierellaceae];g__Parvimonas</th>\n",
       "      <td>2.287149</td>\n",
       "      <td>0.204409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Veillonellaceae;g__Selenomonas</th>\n",
       "      <td>2.300208</td>\n",
       "      <td>0.229052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Lachnospiraceae;g__Catonella</th>\n",
       "      <td>2.401605</td>\n",
       "      <td>0.208750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Erysipelotrichi;o__Erysipelotrichales;f__Erysipelotrichaceae;g__Bulleidia</th>\n",
       "      <td>2.491430</td>\n",
       "      <td>0.186822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Cyanobacteria;c__Chloroplast;o__Streptophyta;f__;g__</th>\n",
       "      <td>2.494325</td>\n",
       "      <td>0.185080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Lachnospiraceae;g__Moryella</th>\n",
       "      <td>2.669319</td>\n",
       "      <td>0.263335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacteriales;f__Enterobacteriaceae;g__Klebsiella</th>\n",
       "      <td>2.969992</td>\n",
       "      <td>0.217478</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        mean       std\n",
       "k__Bacteria;p__Actinobacteria;c__Actinobacteria...  1.092910  0.227672\n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...  1.189515  0.259315\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...  1.221614  0.212661\n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...  1.297860  0.208737\n",
       "k__Bacteria;p__Proteobacteria;c__Betaproteobact...  1.309592  0.211117\n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...  1.389300  0.230333\n",
       "k__Bacteria;p__Bacteroidetes;c__Flavobacteriia;...  1.390022  0.245553\n",
       "k__Bacteria;p__Proteobacteria;c__Betaproteobact...  1.395832  0.190013\n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...  1.533259  0.254360\n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...  1.639519  0.219621\n",
       "k__Bacteria;p__Proteobacteria;c__Betaproteobact...  1.709216  0.159673\n",
       "k__Bacteria;p__Bacteroidetes;c__Flavobacteriia;...  2.121370  0.238828\n",
       "k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o__...  2.255435  0.208298\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...  2.287149  0.204409\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...  2.300208  0.229052\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...  2.401605  0.208750\n",
       "k__Bacteria;p__Firmicutes;c__Erysipelotrichi;o_...  2.491430  0.186822\n",
       "k__Bacteria;p__Cyanobacteria;c__Chloroplast;o__...  2.494325  0.185080\n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...  2.669319  0.263335\n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...  2.969992  0.217478"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "differentials.tail(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another cool thing about this approach is the log-fold change could also be interpreted as a type of feature-importance; the microbes the increased the most and decreased the most are the microbes that best describe the differences between the CD and Control samples.  Furthermore, we can design *classifiers* with this sort of information; by selecting the top microbes and bottom microbes, we can compute their average log ratio per sample which can serve as a 1D classifier given by\n",
    "\n",
    "$$\n",
    "\\log\\bigg(\\frac{g(x_{top} + 1)}{g(x_{bottom} + 1)}\\bigg) \n",
    "$$\n",
    "\n",
    "Where $x_{top}$ corresponds to the abundances of the most increased microbes, $x_{bottom}$ corresponds to the abundances of the most decreased microbes and $g(x)$ is the geometric mean.  Since there are many zero counts, we do need some imputation, so we add a pseudocount to alleviate this issue.  \n",
    "\n",
    "The code to compute this log-ratio across the top 20 microbes and bottom microbes is given below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "numerator = np.log(table.loc[:, differentials.head(20).index] + 1)\n",
    "denominator = np.log(table.loc[:, differentials.tail(20).index] + 1)\n",
    "balance = numerator.mean(axis=1) - denominator.mean(axis=1)\n",
    "md['TopTaxaLogRatio'] = balance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To see how reasonable this log-ratio is, we investigate the distribution of the 3 patient groups for this log-ratio below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:xlabel='diagnosis', ylabel='TopTaxaLogRatio'>"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEGCAYAAABsLkJ6AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAYLElEQVR4nO3dfXRdV33m8e8jS4NNnZfG121K1MQkMjAQnAScF1ZaSggy0RhKeZlpmJn0FmjjzlC7kGFWW6aUkmZNy7TpmiXzUhsauFAopUBWUogaK8VJhtKGOCRx3kisBEPUhMYycZwXJ0jRb/64R0EWtnylq3P3vdrPZy0t6Zx7dc4Ta+X+zt5nn70VEZiZWX66UgcwM7M0XADMzDLlAmBmlikXADOzTLkAmJllqjt1gLmoVCqxatWq1DHMzDrKLbfcMhYRK2fu76gCsGrVKnbs2JE6hplZR5H0vUPtdxeQmVmmXADMzDLlAmBmlikXADOzTLkAmDVpbGyMjRs3snfv3tRRzObEBcCsSbVajZ07d1Kr1VJHMZsTFwCzJoyNjTE0NEREMDQ05FaAdRQXALMm1Go1pqZUn5ycdCvAOooLgFkThoeHGR8fB2B8fJxt27YlTmTWOBcAsyb09/fT09MDQE9PD+vWrUucyKxxLgBmTahWq0gCoKuri2q1mjiRWeOSFQBJSyV9S9Ltku6S9KFUWczmq1KpMDAwgCQGBgZYsWJF6khmDUs5GdwzwGsj4glJPcA3JA1FxL8kzGQ2Z9Vqld27d/vq3zpOsgIQ9aETTxSbPcWXV6i3jlOpVNi8eXPqGGZzlvQegKQlkm4DHgGGI+KmQ7znYkk7JO3Ys2dPyzOamS1WSQtARDwbEacDvcBZkk49xHu2RsTaiFi7cuVPrGdgZmbz1BajgCJiH3A9cEHaJGZm+Ug5CmilpGOLn5cBrwO+kyqPmVluUo4C+jmgJmkJ9UL0xYj4asI8ZmZZSTkKaCdwRqrzm5nlri3uAZiZWeu5AJiZZcoFwMwsUy4AZmaZcgEwM8uUC4CZWaZcAMzMMuUCYGaWKRcAM7NMuQCYmWXKBcDMLFMuAGZmmXIBMDPLlAuAmVmmXADMzDLlAmBmlikXADOzTLkAmJllygXAzCxTLgBmZplyATAzy1R36gBmrTI4OMjIyMiCH3d0dBSA3t7eBT82QF9fH5s2bSrl2JY3FwCzJh04cCB1BLN5cQGwbJR1FT113MHBwVKOb1YW3wMwM8uUC4CZWaZcAMwsa2NjY2zcuJG9e/emjtJyLgBmlrVarcbOnTup1Wqpo7ScC4CZZWtsbIyhoSEigqGhoexaAS4AZpatWq1GRAAwOTmZXSvABaBN5NwPaZbK8PAw4+PjAIyPj7Nt27bEiVrLBaBN5NwPaZZKf38/3d31x6G6u7tZt25d4kStlawASPp5Sdsl3SPpLkm/kypLarn3Q5qlUq1WmZycBOpdQNVqNXGi1krZApgA/kdE/HvgHODdkl6aME8yufdDmlkayQpARDwcEd8ufn4cuAc4IVWelHLvhzRLpVar0dVV/xjs6urK7uKrLe4BSFoFnAHcdIjXLpa0Q9KOPXv2tDxbK/T399PT0wNAT09Pdv2QZqkMDw8zMTEBwMTERHYXX8kLgKTlwJeB90TE/pmvR8TWiFgbEWtXrlzZ+oAtML3fUVJ2/ZBmqeR+8ZW0AEjqof7h/7mI+ErKLClVKhVOOKHe+/WCF7yAFStWJE5klodqtYokoN4FlNvFV8pRQAL+CrgnIv4iVY52MDY2xkMPPQTAQw895FFAZi1SqVQYGBhAEgMDA9ldfKVsAZwLXAS8VtJtxdd/SJgnmemjgCIiuxtRZilVq1XWrFmT3dU/pB0F9I2IUESsiYjTi69rUuVJyaOAzNKpVCps3rw5u6t/aIObwOYbUWaWhgtAG8j9RpSZpeEC0AZyvxFlZml4Ufg2Ua1W2b17t6/+zaxl3AIwM8uUC0Cb8HTQZtZqLgBtwNNBm1kKRywAknokbZL0peJrYzGFgy0QTwdtZik00gL4OPBK4GPF1yuKfbZA/CCYmaXQyCigMyPitGnbX5d0e1mBctTf388111zD+Pi4HwQzO4zBwUFGRkYW/Lijo6MA9Pb2Lvix+/r62LRp04Ifd6E00gJ4VtIpUxuSTgaeLS9SfvwgmFk6Bw4c4MCBA6ljJNFIC+B/AtslPQAIOAl4R6mpMlOpVDjvvPO49tprOe+88/wgmNkhlHUlPXXcwcHBUo7fzo5YACLiHyWtBl5MvQB8JyKeKT2ZmZmV6rBdQJJeW3x/C7Ae6ANOAdYX+2yBjI2NsX37dgC2b9/uYaBm1hKz3QP4peL7Gw/x9YaSc2XFw0DNLIXDdgFFxAeLHy+NiO9Of03SC0tNlZlDDQO95JJLEqcys8WukVFAXz7Evi8tdJCceT0AM0vhsC0ASS8BXgYcM6PP/2hgadnBclKtVhkaGgI8DNTMWme2FsCLqff1H8vB/f+vAH6z9GQZ8XoAZpbCbPcArgKukvSqiPjnFmbKktcDMLNWa+RBsFslvZt6d9BzXT8R8c7SUmVoamFqM7NWaeQm8GeB44HXAzcAvcDjZYYyM7PyNVIA+iLiA8CTEVGj/lDYy8uNZWZmZWukC2i8+L5P0qnAD4BVpSWy7JU162NZdu3aBZQ3V01Z2n2mSitfIwVgq6SfBv4AuBpYDnyg1FSWtZGREe6789ucuLwzJp39d+P1hvTTu29OnKRx339iSeoI1gYamQzuk8WPNwInA0g6qcxQZicuf5Y/WPtE6hiL1mU7lqeOYG1g1nsAkl4l6W2SfqbYXiPp88A3WpLOzMxKM9tsoH8GXAG8FfiapA8Cw8BNwOrWxDMzs7LM1gW0HjgjIp4u7gE8BKyJiF2tiWZmZmWarQAciIinASLiUUn35v7hX+bolJzXJTWzNGYrAKdIunra9qrp2xHxy+XFyk+ua5KaWTqzFYA3zdi+vMwgnaDMq+ic1yU1szRmmwzuhrJPLukK6jOOPhIRp5Z9PjMz+7EjPgcg6Q4gZux+DNgBXBYRzSxg+2ngI8BnmjiGmZnNQyNPAg8BzwKfL7YvBES9CHya+hoB8xIRN0paNd/fNzOz+WukAJwbEedO275D0j9FxLmS/mtZwaZIuhi4GODEE08s+3Rm1oROm8cJ8p7LqZECsFzS2RFxE4Cks6jPBwQw0dTZGxARW4GtAGvXrp3ZFWVmbWRkZIRb77q1vo5gp5isf7v1X29Nm2Mu9i3MYRopAL8BXCFpOfWun/3AuyT9FPAnCxPD7MdGR0d58vElnq+mRN97fAk/VTx7suCOhcnXTJZzbAOg6/pGZvI/skYmg7sZeLmkYwBFxL5pL39xQVKYmVnLNTIK6Bjgg8Cri+0bgEsj4rFmTy7pb4DXABVJo8AHI+Kvmj2udbbe3l6ennjYs4GW6LIdy1lawlPn1lka6QK6ArgT+E/F9kXAp4C3NHvyiHh7s8cwM7P5aaQAnBIRb522/SFJt5WUx8zMWqSROwkHJP3C1IakcwFPXGNm1uEaaQH8FvCZ4l4AwKNAtbxIZmbWCo2MArodOE3S0cX2fknvAXaWnM3MzErU8GDSiNgfEfuLzUtKymNmZi0y36cJtKApzMys5eZbADwlg5lZhzvsPQBJj3PoD3oBy0pLZGYda3R0FB5buKkK7DD2wWg0P5XHbAvCHNX00c3MrG01MgwUAEk/Ayyd2o6I75eSyMw6Vm9vL3u0x5PBlazr+i56T2h+Ko8jttMk/bKkXcB3gRuA3dQXiTEzsw7WSEfdHwPnAPdFxAuB84F/KjWVmZmVrpECMF6s+9slqSsitgOnlxvLzMzK1sg9gH3FYjA3Ap+T9AgtWAnM8vb9JzpnQZh/e6p+HfWzz++cfu/vP7GEF6UOYck1UgDeBDwNvBf4L8AxwKVlhrK89fX1pY4wJz8q1pRdump14iSNexGd9+9sC6+RAnBSRNxd/FwDkPQa4PpyIlnuOm1x7qm8g4ODiZOYzU0j9wC+KOl3VbdM0ma8FrCZWcdrpACcDfw88E3gZuAh4NwyQ5mZWfkaGgVEfQGYZdQfBPtuRHTO3S4zMzukRgrAzdQLwJnALwBvl/SlUlOZmVnpGrkJ/K6I2FH8/APgTZIuKjGTmXWyfR02GdwTxffOGHVctw84ofnDNLIi2A74ibmAbmj+1Ga22HTi0NJdxTDe1Sd0zjBeTliYf+sjFgBJbwT+AngB8AhwEnAP8LKmz25mi0qnDeGFvIfxNtJOuwzPBWRmtug0cg9gPCL2SnpuLiBJHy49WRMGBwcZGRlJHWNOppqhnXQF1dfX11F5zexgs60I9tsR8RE6cC6gkZERbr3jbiaff1zqKA3Tj+qLr91y/w8SJ2lM11M/TB3BzJo0WwvgncBH6NC5gCaffxxPv/QNqWMsWkvv/mrqCGbWpEZGAT05bbNWYhYzM2uh2QrAGkn7D7FfQETE0SVlMjOzFpitANwREWe0LImZmbVUBz2uZ2ZmC2m2AvB3LUthZmYtd9gCEBH/G0DSyZL+XtKYpEckXSXp5IU4uaQLJN0raUTS7y3EMc3MrDGNdAF9HvgicDz16SD+DvibZk8saQnwUWAAeCn1WUZf2uxxzcysMY0UAEXEZyNiovj6ayAW4NxnASMR8UBE/Aj4AvVnDszMrAUamQpie9E98wXqH/y/CnxN0nEAETHfR0JPAB6ctj1KffWxg0i6GLgY4MQTT2zowKOjo3Q99ZgfVipR11N7GR1t6wfCzewIGikAv1p83zBj/zupF4T53g/QIfb9RMsiIrYCWwHWrl27EC0PMzOjsSeBX1jSuUeprzU8pZf6esNN6+3t5d+e6fZUECVaevdX6e09PnUMM2tCI+sB9AD/DXh1set6YEtEjDd57puB1ZJeCPwrcCHwn5s8ppmZNaiRLqCPAz3Ax4rti4p9v9HMiSNiQtJvA9cCS4ArIuKuZo5pZmaNm2066O6ImADOjIjTpr30dUm3L8TJI+Ia4JqFOJaZmc3NbMNAv1V8f1bSKVM7i4fAni01lZmZlW62LqCpUTrvoz4U9IFiexXwjjJDmZlZ+WYrACslXVL8vIV6P/2TwFLgDGB7ydnMzKxEsxWAJcByDh6vv7z4flRpiczMrCVmKwAPR0TbL/1oZmbzM9tN4EM9qWtmZovEbAXg/JalMDOzlpttPYD5TvJmZmYdoJEngc0WhcHBQUZGRhb8uLt27QJg06ZNC35sgL6+vtKObXlzATBr0rJly1JHMJsXFwDLhq+izQ62aAtA11M/7KgFYfT0fgBi6dGJkzSm66kfUl8l1Mw61aIsAH19fakjzNmuXY8DsPqUTvlQPb4j/53N7McWZQHoxKb+VObBwcHEScwsF40sCm9mZouQC4CZWaZcAMzMMuUCYNaksbExNm7cyN69e1NHMZsTFwCzJm3ZsoXbb7+dLVu2pI5iNieLchSQWauMjY0xPDwMwLZt29iwYQMrVqxInGpx6sSpPNp9Gg+3AMyasGXLFiYnJwGYnJx0K6ADLVu2LNvpPNwCMGvCddddd9D28PAw73//+xOlWdza+Uq6U7kFYNYESbNum7UzFwCzJpx//sHrJr3uda9LlMRs7lwAzJqwYcMGurrq/xt1dXWxYcOGxInMGucCYNaESqXC2WefDcA555zjEUDWUVwAzJo0OjoKwIMPPpg4idncuACYNeG+++577oP/wQcfLGWcullZXADMmnDZZZcdtH3ppZcmSmI2dy4AZk3YvXv3rNtm7cwFwKwJy5cvn3XbrJ25AJg1YWJiYtZta385z+aapABI+o+S7pI0KWltigxmC+H1r3/9QdsXXHBBoiQ2X7VajZ07d1Kr1VJHablULYA7gbcANyY6v9mCqFardHfXp9Tq6emhWq0mTmRzMTY2xtDQEBHB0NBQdq2AJAUgIu6JiHtTnNtsIVUqFdavX48k1q9f7wfBOkytViMigPpsrrm1Atr+HoCkiyXtkLRjz549qeOY/YRqtcqaNWt89d+BhoeHGR8fB2B8fJxt27YlTtRapRUASddJuvMQX2+ay3EiYmtErI2ItStXriwrrtm8VSoVNm/e7Kv/DtTf309PTw9Q78Jbt25d4kStVdp6ABHhaRHNrK1Vq1WGhoaA+mR+ubXi2r4LyMysLJVKhYGBASQxMDCQXSsu1TDQN0saBV4FfE3StSlymJnlfA8nyZKQEXElcGWKc5uZTTd1DydH7gIyM8uUC4CZWaZcAMzMMuUCYGaWKRcAM7NMuQCYmWXKBcDMLFMuAGZmmXIBMDPLlAuAmVmmXADMzDLlAmBmlqkkk8F1qsHBQUZGRko59q5duwDYtGnTgh+7r6+vlOOaWWdzC6BNPO95z+OZZ555bnk6M7OyuQUwB2VeRV9++eVcffXVrF69mksuuaS085iZTXELoA2MjY0xNDRERDA0NMTevXtTRzKzDLgAtIFarUZEADA5OUmtVkucyMxy4ALQBoaHh5/r+x8fH2fbtm2JE5lZDlwA2kB/fz/d3fXbMd3d3axbty5xIjPLgQtAG6hWq0xOTgL1LqAcF6c2s9ZzATAzy5QLQBuo1Wp0ddX/FF1dXb4JbGYt4QLQBoaHh5mYmABgYmLCN4HNrCVcANpAf38/PT09APT09PgmsJm1hAtAG6hWq0gC6l1AvglsZq3gAtAGKpUKAwMDSGJgYIAVK1akjmRmGfBcQG2iWq2ye/duX/2bWcu4ALSJSqXC5s2bU8cws4y4C8jMLFMuAGZmmXIBMDPLlAuAmVmmNDUPfSeQtAf4XuocJaoAY6lD2Lz4b9fZFvvf76SIWDlzZ0cVgMVO0o6IWJs6h82d/3adLde/n7uAzMwy5QJgZpYpF4D2sjV1AJs3/+06W5Z/P98DMDPLlFsAZmaZcgEwM8uUC0ACko6X9AVJ90u6W9I1kl4k6YCkWyXdI+lbkjw1qNkCkrRK0p0z9v2RpPcVP79P0nck3Snpdkm/liZpa3g20BZTfeWXK4FaRFxY7Dsd+Fng/og4o9h3MvAVSV0R8alUec1yIem3gH7grIjYL+kY4FfSpiqXWwCtdx4wHhF/ObUjIm4DHpz+poh4ALgE2NTSdHZExVXkPZI+IekuSdskLZN0uqR/kbRT0pWSfjp1VpuT9wP/PSL2A0TEYxFRS5ypVC4ArXcqcEuD7/028JISs9j8rQY+GhEvA/YBbwU+A/xuRKwB7gA+mC6ezdEy4KiIuD91kFZyAWhvSh3ADuu7RcsN6gX9FODYiLih2FcDXp0imM3qcOPeu2Z5bdFyAWi9u4BXNvjeM4B7Ssxi8/fMtJ+fBY5NlMPmZi8ws2vuOGA38GRx7y0bLgCt93XgeZJ+c2qHpDOBk6a/SdIq4M8BrxPZGR4DHpX0i8X2RcANs7zfEoiIJ4CHJZ0PIOk44ALgG8CfAB+VdHTx2tGSLk4WtgU8CqjFIiIkvRn4v5J+D3ia+tXHe4BTJN0KLAUeBzZ7BFBHqQJ/Ken5wAPAOxLnsUP7Neof9JcX2x+KiPslfRxYDtwsaRwYBy4/3EEWA08FYWaWKXcBmZllygXAzCxTLgBmZplyATAzy5QLgJlZpjwM1LIl6Y+AJ4CjgRsj4rqEWS5NncHy4wJg2YuIP3QGy5G7gCwrkv6XpHslXQe8uNj3aUlvK37+Q0k3F/PBby2m70bSmcUsn/8s6c+m5pSX9OuSviLpHyTtkvR/pp3r7ZLuKI714WLfkuJ8dxavvfcQGf60WCdip6Q/b+k/kGXFLQDLhqRXAhdSn2Opm/psqzNnZv1IRFxavP+zwBuAvwc+BVwcEd+U9Kczfuf04pjPAPdK2kx9fqAPU5/36VFgm6RfoT7t9wkRcWpxjmNnZDwOeDPwkuKp8YNeN1tIbgFYTn4RuDIinirmfL/6EO85T9JNku4AXgu8rPgQPioivlm85/Mzfucfi7njnwbupj6v05nA9RGxJyImgM9Rnx30AeBkSZslXQDsn3Gs/dSnB/mkpLcATzX7H212OC4AlpvDzn0iaSnwMeBtEfFy4BPU52U60rTcM2cG7T7c70TEo8BpwPXAu4FPznh9AjgL+DL11aj+4QjnNps3FwDLyY3Am4vVu44C3jjj9aXF9zFJy4G3wXMf2o9LOqd4/cIGznUT8EuSKpKWAG8HbpBUAboi4svAB4BXTP+l4rzHRMQ11CcIPH2O/41mDfM9AMtGRHxb0t8CtwHfA/7fjNf3SfoE9dW8dgM3T3v5XcAnJD1J/er9sSOc62FJvw9sp94auCYirpJ0GvApSVMXX78/41ePAq4qWiMC3jvX/06zRnk2ULMGSFpezCVPMY33z0XE7ySOZdYUtwDMGrO+uKLvpt56+PW0ccya5xaAmVmmfBPYzCxTLgBmZplyATAzy5QLgJlZplwAzMwy9f8BXM8aM4xpZ6oAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "sns.boxplot(x='diagnosis', y='TopTaxaLogRatio', data=md)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see, there is a huge separation between the Control group (no) and the CD group.  We can further validate this by calculating the AUC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8091900311526479"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn import metrics\n",
    "truth = md['diagnosis'].values == 'no'\n",
    "pred = balance.values\n",
    "fpr, tpr, thresholds = metrics.roc_curve(truth, pred)\n",
    "metrics.auc(fpr, tpr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One interesting thing about this is the training AUC is slightly better than the AUC of 0.78 that was obtained using Random Forests in the original dataset.  And this approach only used 1 dimension!\n",
    "\n",
    "Before we get too excited, we do need to point out the potential caveats.  We aren't performing cross-validation, so there is always the possibility that we are overfitting. If we wanted to be rigorious about this approach, we would want to do careful cross-validation, where we train our differential abundance on a subset of the dataset, and evaluate our classification accuracy on a hold out dataset.\n",
    "\n",
    "Indeed we can and should do this!  This tutorial is designed to give a sense of the additional possibilities surrounding differential abundance analysis that could be exercised."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusion\n",
    "\n",
    "There is a huge world of possibilities surrounding differential abundance analysis that can be realized with insights from Compositional Data Analysis.  It is important to note that this is just the tip of the iceberg. There are many more things that can be discussed in the context of correlations, dimensionality reduction, classification and much more!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
